implement+test mean for datetimelike EA/Index/Series

jbrockmendel opened the pull request:
This takes over from #23890, implementing+testing reductions one at a time.

- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

jbrockmendel committed:
implement+test mean for datetimelike EA/Index/Series

gfyoung commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -1434,6 +1434,34 @@ def max(self, axis=None, skipna=True, *args, **kwargs):
         # Don't have to worry about NA `result`, since no NA went in.
         return self._box_func(result)
 
+    def mean(self, axis=None, skipna=True):
```
List of parameters?

jreback commented:
we are still treating datetimes as non-numeric yes? meaning that by default they will be exlcuded from say .mean() and you have to pass ``numeric_only=False``?

jreback reviewed (changes requested):
can you see if we have coverage of this on mixed dtype frames as well?  (see my comment about numeric_only). Also need to assert that other ops automatically exclude (e.g. std).

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel committed:
update imports

jbrockmendel committed:
isort fixup

jreback commented on pandas/core/series.py:
```diff
@@ -3614,6 +3614,9 @@ def _reduce(self, op, name, axis=0, skipna=True, numeric_only=None,
         elif is_datetime64_dtype(delegate):
             # use DatetimeIndex implementation to handle skipna correctly
             delegate = DatetimeIndex(delegate)
+        elif is_timedelta64_dtype(delegate) and hasattr(TimedeltaIndex, name):
```
why would you need to check if the method exists we don't do this for any other types

jbrockmendel replied:
otherwise this breaks for e.g. `std`

jreback replied:
but it doesn't break for datetime64, so puzzled why this is needed here

jreback replied:
?

jbrockmendel replied:
not sure what you're asking for here, buddy.

jreback replied:
eel my comment is why are you checking if the method exists when we don’t use this anywhere else

jbrockmendel replied:
because failing to check for this causes a test failure.  In particular, for `std`, where a timedelta64 ndarray has the method but TDA doesn't yet

jreback replied:
the is true for DTA so your answer doesn’t make sense

jbrockmendel replied:
`std` is _supposed_ to raise for datetime64 in the status quo, so we don't test it

jreback replied:
still this is just smelly. I don't see a problem delegating to TDI here if its a timedelta type. there must be something else wrong if you need this check.

jreback commented on pandas/tests/reductions/test_stat_reductions.py:
```diff
@@ -11,9 +11,74 @@
+        # shuffle so that we are not just working with monotone-increasing
+        dti = dti.take([4, 1, 3, 10, 9, 7, 8, 5, 0, 2, 6])
+
+        # use hourly frequency to avoid rounding errors in expected results
```
not sure mean actually should work on PeriodIndex/Array

jbrockmendel replied:
how come?  the meaning is unambiguous

jreback replied:
I guess these have to be uniform periods so mean is just another period, though I think this could easily be a non-sensical return values, e.g. what is the mean of 

```
In [2]: pd.date_range('2012', freq='M', periods=4)                                                                                                                                                                                                                      
Out[2]: DatetimeIndex(['2012-01-31', '2012-02-29', '2012-03-31', '2012-04-30'], dtype='datetime64[ns]', freq='M')

```

jorisvandenbossche replied:
>  the meaning is unambiguous

Can you explain that more? Maybe from an implementation point (viewing it as integers) it is clear what to do, but semantically for periods, I certainly don't find this obvious.

jreback reviewed (changes requested):
need the tests from the OP (e.g. a mixed DataFrame)

TomAugspurger commented:
> we are still treating datetimes as non-numeric yes? meaning that by default they will be exlcuded from say .mean() and you have to pass numeric_only=False?

I think that makes the most sense.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel committed:
params for docstring

jbrockmendel committed:
test for numeric_only

TomAugspurger commented:
Does this close https://github.com/pandas-dev/pandas/issues/24752?

jbrockmendel commented:
No, no it does not.

jreback commented on pandas/tests/indexes/datetimes/test_ops.py:
```diff
@@ -47,6 +47,27 @@ def test_ops_properties_basic(self):
         pytest.raises(AttributeError, lambda: s.weekday)
 
+    # TODO: figure out where in tests.reductions this belongs
+    def test_mean(self, tz_naive_fixture):
```
seems *very* duplicative of the test below, is there a reason?

jbrockmendel replied:
bc at the time when this was written Index reduction tests had not been collected in tests.reductions.  Will move/de-duplicate.

jbrockmendel commented:
> I guess these have to be uniform periods so mean is just another period, though I think this could easily be a non-sensical return values, e.g. what is the mean of `pd.date_range('2012', freq='M', periods=4)`

Eyeballing it at about 2012-03-14.  Why would that be nonsensical?

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jreback commented:
it’s no longer a M freq Period. my point is the mean if a span is an odd thing

jorisvandenbossche commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -1435,6 +1435,41 @@ def max(self, axis=None, skipna=True, *args, **kwargs):
 
+    def mean(self, axis=None, skipna=True):
+        """
+        Return the mean value of the Array or maximum along
```
Why "or maximum" ?

jbrockmendel replied:
copy/paste screwup, will fix

jbrockmendel committed:
copy/paste fixup

jbrockmendel commented:
> Can you explain that more? Maybe from an implementation point (viewing it as integers) it is clear what to do, but semantically for periods, I certainly don't find this obvious.

Period/PeriodArray/PeriodIndex is the closest thing we have to "lower-resolution Timestamp/DTA/DTI".  In this context, the mean of a bunch of Periods is analogous to a mean of a bunch of Timestamps.

jbrockmendel commented:
> it’s no longer a M freq Period. my point is the mean if a span is an odd thing

@jreback in the example you gave, it isn't a Period at all.  It is a collection of Timestamps that happen to be a month apart.

TomAugspurger commented:
I'm guessing @jreback's example was a typo and meant a `period_range`.

I'm still wrapping my head around the meaning of a mean for period data. The thing tripping me up is the necessary "rounding" involved in the mean. I think a good analogy is to integers.

```python
In [14]: np.arange(12).mean()
Out[14]: 5.5

In [15]: pd.period_range('2000', periods=12, freq='M').mean()
Out[15]: Period('2000-06', 'M')
```

so the `mean[int] -> float`.

It's not obvious to me that preserving the type and rounding is preferable to type casting (to a  Timestamp). But I don't think returning a Timestamp is feasible because then (IIUC) you have to worry about weird things like length of months, etc.

So the only real option is returning a Period that's rounded to the nearest period for our freq. But it took me a decent amount of thinking to get to that conclusion, and I'm still not sure it's obviously correct. For now, I'd recommend disabling mean for Period, unless we have a compelling reason to support it.

jorisvandenbossche commented:
> Period/PeriodArray/PeriodIndex is the closest thing we have to "lower-resolution Timestamp/DTA/DTI". In this context, the mean of a bunch of Periods is analogous to a mean of a bunch of Timestamps.

It might be the closest thing, but it is not a lower-resolution timestamp. It represents a time span, and I don't think it is that clear what the mean of several time spans should be.

>  For now, I'd recommend disabling mean for Period, unless we have a compelling reason to support it.

\+1

jreback commented:
could provide a nice NotImplememtedError saying to convert to time stamps first

jbrockmendel commented:
@jreback can you confirm Tom's hypothesis that you meant period_range instead of date_range?  i.e. are we all OK with `DatetimeArray.mean`?

For PeriodArray I agree that for non-Tick frequencies it gets wonky.  Special-casing Tick vs non-Tick would be a hassle I'm OK with avoiding.  So will disable this for PeriodArray.

jreback commented:
yes was a typo

ok on DTA / TDA but i think consensus is to not try to handle PA array currently

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel committed:
Disable for PeriodArray

jbrockmendel committed:
Delete assertions missed in previous commit

jreback commented on pandas/tests/frame/test_analytics.py:
```diff
@@ -1304,6 +1304,39 @@ def test_mean_corner(self, float_frame, float_string_frame):
+        })
+        tm.assert_series_equal(result, expected)
+
+        # FIXME: df.mean(numeric_only=False) raises TypeError because
```
don't comment a test, rather xfail it

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel committed:
xfail numeric_only=False case

jbrockmendel committed:
Merge branch 'means' of https://github.com/jbrockmendel/pandas into means

jreback reviewed (changes requested):
see comments

jbrockmendel commented:
> I don't see a problem delegating to TDI here if its a timedelta type. there must be something else wrong if you need this check.

The problem is that TDI/TDA doesn't have `std`, `sum`, etc yet, while ndarray[m8[ns]] does.

@TomAugspurger am I not communicating this clearly?

jbrockmendel commented:
@jorisvandenbossche when convenient, can I get your help clearing up an apparent communication problem?

TomAugspurger commented:
Sorry, catching up on emails...

Making sure I understand the PR correctly. This ensure that Series/Index .mean works for DatetimeArray/TimedeltaArray, while not breaking `.str` for DTA/DTA-backed Series/Index?.

If you didn't include the `and hasattr(TimedeltaIndex, name)`, then we'd end up calling `TimedeltaIndex(...).std()`, which fails, because TimedeltaIndex.std isn't a method?

IOW, this works on master

```python
In [2]: arr = pd.array([1, 2, 3], dtype='timedelta64[ns]')

In [3]: pd.Series(arr).std()
Out[3]: Timedelta('0 days 00:00:00.000000')
```

but would break without the `and hasattr(TimedeltaIndex, name)`?

jbrockmendel commented:
> [...] but would break without the and hasattr(TimedeltaIndex, name)?

Right.

TomAugspurger commented:
Thanks,  once all of the ops that we want are implemented, the `and hasattr(TimedeltaIndex, name)` can be removed?

Does that make sense @jreback? We can't exclude that check until TDA has all the methods, and we probably prefer smaller PRs?

jreback commented:
why don't we add NotImplented to those ops then?

jbrockmendel commented:
> why don't we add NotImplementedErr to those ops then?

For starters because that would also cause the relevant tests to fail, just slightly differently.

Second because after this PR goes through the next one will implement std, followed by median and whatever else is relevant.

jreback commented:
ok can we do one of 2 things here:

- remove Timedelta change for now until we don't need to add this accessor check
- leave the change, but add a big TODO (and followup as soon as possible)?

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel committed:
add todo comment

jbrockmendel commented:
added TODO comment

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel committed:
dont expect datetime in dataframe.mean

jorisvandenbossche reviewed (changes requested):
Given the discussion we had on the other issue (where we disallowed M8 again for mean on Series), I am actually not sure that we should implement `DatetimeArray.mean`

jreback commented:
> Is there a reason why DatetimeArray.mean would not be meaningful or have an ambiguous meaning? If not, why wouldn't you want the method to exist?

I am with @jbrockmendel here, DTA is well defined an meaningful, PeriodArray.mean is problematic which is the other discussion

jorisvandenbossche commented:
> Is there a reason why DatetimeArray.mean would not be meaningful or have an ambiguous meaning? If not, why wouldn't you want the method to exist?

Do you know if there is a reason why numpy choose to not support it?

I find the argument of Tom a convincing one: a mean of integers is a float, which can loose precision / give rounding. Altough, the same is true for timedelta of course ..

TomAugspurger commented:
Ah I didn't realize that timedelta did implement it. Looking at the exception

```python
In [5]: np.mean(np.array([1, 2], dtype='M8[ns]'))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-343ebecd9dd6> in <module>
----> 1 np.mean(np.array([1, 2], dtype='M8[ns]'))

~/Envs/pandas-dev/lib/python3.7/site-packages/numpy/core/fromnumeric.py in mean(a, axis, dtype, out, keepdims)
   3116
   3117     return _methods._mean(a, axis=axis, dtype=dtype,
-> 3118                           out=out, **kwargs)
   3119
   3120

~/Envs/pandas-dev/lib/python3.7/site-packages/numpy/core/_methods.py in _mean(a, axis, dtype, out, keepdims)
     73             is_float16_result = True
     74
---> 75     ret = umr_sum(arr, axis, dtype, out, keepdims)
     76     if isinstance(ret, mu.ndarray):
     77         ret = um.true_divide(

TypeError: ufunc add cannot use operands with types dtype('<M8[ns]') and dtype('<M8[ns]')
```

Not supporting datetimes is likely not a deliberate choice one way or the other, just a consequence of not having sum for datetimes.

jorisvandenbossche commented:
Yes, that might be true. Also given a comment of @shoyer elsewhere (https://github.com/pandas-dev/pandas/issues/9442), historically this might indeed been simply not implemented, instead of explicitly disallowed. @shoyer is that correct?

In any case, currently our datetime64 mean support is very inconsistent: eg DataFrame and Series raise an error (also if you do `numeric_only=False`), but in groupby we do support it (which I learned from Jeff here on [SO](https://stackoverflow.com/questions/24328834/groupby-and-average-datetime64) :))

shoyer commented:
Yes, I don't think the omission of mean for datetime64 was intentional for numpy, but fixing this the "right" way upstream in numpy will be challenging -- see https://github.com/numpy/numpy/issues/12901

jbrockmendel commented:
> What's the expected value of `np.mean([pd.Timestmap(1), pd.Timestamp(2)])`

`pd.Timestamp(1.5)`

Rounding error is a fact of life that anyone working with nanosecond-precision data will be aware of.  I could concoct an example with int64 data that was lossy, but that doesn't mean we shouldn't support mean for int64.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jreback commented:
Yeah we fully support this for timedelta, I don't see why we don't support for M8[ns] as well.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel commented:
@jorisvandenbossche gentle ping; this isn't technically a blocker, but I'd like to get it in before moving on to the other reductions.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel commented:
@jorisvandenbossche gentle ping

jbrockmendel commented:
@jreback how to move forward here?

jreback commented:
@jbrockmendel pls add a whatsnew note. This is fine as it sidesteps the issue with PeriodArray where things are ambiguous. Timestamp / DTI are not and worthy of this.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel committed:
whatsnew

jreback commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -1435,6 +1435,48 @@ def max(self, axis=None, skipna=True, *args, **kwargs):
+            Dummy parameter to match NumPy signature
+        skipna : bool, default True
+            Whether to ignore any NaT elements
+
```
can you add a versionadded

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -24,6 +24,7 @@ Other Enhancements
 - :meth:`DataFrame.at_time` and :meth:`Series.at_time` now support :meth:`datetime.time` objects with timezones (:issue:`24043`)
 - :meth:`DataFrame.set_index` now works for instances of ``abc.Iterator``, provided their output is of the same length as the calling frame (:issue:`22484`, :issue:`24984`)
 - :meth:`DatetimeIndex.union` now supports the ``sort`` argument. The behaviour of the sort parameter matches that of :meth:`Index.union` (:issue:`24994`)
+- :class:`DatetimeIndex` and :class:`TimedeltaIndex` now have a `mean` method (:issue:`24757`)
```
can you add in the api/references

jbrockmendel replied:
is that in this file or somewhere else?

TomAugspurger replied:
You would need to

1. Add to the class docstring in DatetimeIndex and TimedeltaIndex under the methods section
2. Add to the lists in doc/source/reference/indexing.rst

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel committed:
add versionadded

jorisvandenbossche commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -1435,6 +1435,50 @@ def max(self, axis=None, skipna=True, *args, **kwargs):
         # Don't have to worry about NA `result`, since no NA went in.
         return self._box_func(result)
 
+    def mean(self, axis=None, skipna=True):
```
Do we want to add an `axis` argument here? 
I know we have already added in some places, but also not added in other places, so probably something we should discuss.

I would personally try to not add too much overhead in additional useless kwargs just for matching signatures, *if* they can be avoided.

TomAugspurger replied:
IIRC, this lets us hook into `np.mean`, but we should verify exactly which kwargs are necessary.

jorisvandenbossche replied:
Yes, that's why we did it in the past (but for that we also need the additional kwargs). But, for EAs, we could also decide to implement the array function protocol to achieve this goal, which might obviate the need to add the additional kwargs

jreback replied:
@jbrockmendel i think its worthile here to allow calling ``np.mean``

jreback replied:
don'e we also need out=? meaning we should just accept kwargs like we do for Series.mean

jorisvandenbossche commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -1435,6 +1435,50 @@ def max(self, axis=None, skipna=True, *args, **kwargs):
+        Parameters
+        ----------
+        axis : None
+            Dummy parameter to match NumPy signature
```
If it is to match numpy's signature, then there are a bunch of other kwargs to add as well

jreback replied:
right and we do this elsewhere

jreback replied:
would just add kwargs for this purpose; we don't add them explicity anywhere else

jorisvandenbossche commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -1435,6 +1435,50 @@ def max(self, axis=None, skipna=True, *args, **kwargs):
+        Return the mean value of the Array or mean along an axis.
+
+        .. versionadded:: 0.25.0
+
```
Can you add here that this is not implemented for PeriodArray?

jorisvandenbossche commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -1435,6 +1435,50 @@ def max(self, axis=None, skipna=True, *args, **kwargs):
+        """
+        if is_period_dtype(self):
+            # See discussion in GH#24757
+            raise NotImplementedError(
```
NotImplementedError seems to indicate that we just not *yet* implemented it, while here we actually explicitly don't want to calculate it. Would a TypeError be more suitable?

jorisvandenbossche reviewed (commented):
@jbrockmendel sorry for the slow reply. Given the further discussion and the comment of @shoyer, I am fine with adding that ability.

I have some further comments on the implementations though. In addition to the inline ones: currently we have no Index class that has a `mean` method, not even the numerical ones. So if we add it here, I would like to see a more consistent solution for it (or at least the idea to do it). Or otherwise for now just add it to the Array class. But in any case, we should revisit the numerical ops on Index classes.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel committed:
change NotImplementedError to TypeError

jbrockmendel committed:
add mean to class docstrings and docs/source/reference/indexing.rst

jbrockmendel commented:
Just pushed, addressing most of the comments:
- changed NotImplementedError to TypeError per @jorisvandenbossche 's request.  Also added a note in the docstring that it is not defined for Period.
- created `Methods` sections for DatetimeIndex and TimedeltaIndex in doc/source/reference/indexing.rst and added `mean` to each of these.

Did not do anything with the kwargs

jreback reviewed (changes requested):
lgtm. should add suppport for np.mean (other kwargs)

TomAugspurger commented:
I think that the only outstanding issue here is how `np.mean(DTA)` behaves. Adding a test (and any necessary dummy kwargs) ensuring that it works would be great.

jbrockmendel commented:
Isn't `np.mean(DTA)` handled by `__array_something__`?

TomAugspurger commented:
I don't think `mean` is a ufunc. I believe numpy just dispatches directly to the `.mean` method if the object has one (best to verify).

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel commented:
> I don't think mean is a ufunc. I believe numpy just dispatches directly to the .mean method if the object has one (best to verify).

At least in numpy 1.16.2 it looks like a ufunc:

```
>>> dti = pd.date_range('2019-03-13', periods=3)
>>> np.mean(dti)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "lib/python2.7/site-packages/numpy/core/fromnumeric.py", line 3118, in mean
    out=out, **kwargs)
  File "lib/python2.7/site-packages/numpy/core/_methods.py", line 75, in _mean
    ret = umr_sum(arr, axis, dtype, out, keepdims)
TypeError: ufunc add cannot use operands with types dtype('<M8[ns]') and dtype('<M8[ns]')
```

jreback commented:
can you merge master. a few comments IIRC.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel commented:
Merged master.  The comments were about np.mean, which does not work on this branch ATM.  But it also doesn't work on master, so that doesn't _have_ to be a blocker.

jorisvandenbossche commented:
> At least in numpy 1.16.2 it looks like a ufunc:

It's for sure not a ufunc:

```
In [23]: type(np.sin)                                                                                                                                                                                               
Out[23]: numpy.ufunc

In [24]: type(np.mean)                                                                                                                                                                                              
Out[24]: function
```

but I assume that the `np.mean` function might be implemented in terms of ufuncs (you see a message regarding "ufunc add").

> woops, that was the error I get on master, not on the PR branch. On this branch I get a TypeError complaining about an unexpected kwarg 'dtypebeing passed tonp.mean`. I'll have to dig into this.

That is because on master it is trying the numpy implementation (because there is no method with the name), resulting in saying that addition is not implemented for datetime64. On this PR, it will try to dispatch to the new method, and then complains about non-matching keywords.

>  But it also doesn't work on master, so that doesn't have to be a blocker.

If we don't intent to have it working here (which is totally fine with me), then I would suggest to also remove the `axis` keyword in this PR (which you only added for compatibility with numpy, is that correct?). We can always add it later when we actually fix the numpy compatibility.

jreback commented:
was just about to merge then noticed this comment.

> If we don't intent to have it working here (which is totally fine with me), then I would suggest to also remove the axis keyword in this PR (which you only added for compatibility with numpy, is that correct?). We can always add it later when we actually fix the numpy compatibility.

we have this same signature everywhere else, so this is ok with me.

jorisvandenbossche commented:
> we have this same signature everywhere else, so this is ok with me.

Where do we have this signature? (assuming your "this is ok with me" refers to the current PR, and not to my comment to remove `axis` for now?)

jreback commented:
look at Series.mean

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jorisvandenbossche commented:
> look at Series.mean

Yes, and as I said above, here we validate kwargs in general, and not just `axis`.

jorisvandenbossche commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -1437,6 +1437,54 @@ def max(self, axis=None, skipna=True, *args, **kwargs):
+        if is_period_dtype(self):
+            # See discussion in GH#24757
+            raise TypeError(
+                "mean is not implemented for {cls} since the meaning may be "
```
```suggestion
                "mean is not implemented for {cls} since the meaning is "
```

jbrockmendel replied:
The thing here is that it isn't ambiguous for Tick-frequencies, so "may be" is more accurate.

jorisvandenbossche replied:
What is the unambiguous result of the mean of this?

```
In [12]: idx = pd.period_range('2012-01-01', periods=4, freq='1s')

In [13]: idx
Out[13]: 
PeriodIndex(['2012-01-01 00:00:00', '2012-01-01 00:00:01',
             '2012-01-01 00:00:02', '2012-01-01 00:00:03'],
            dtype='period[S]', freq='S')
```

jbrockmendel replied:
`Period('2012-01-01 00:00:01', freq='S')`.  We already discussed rounding error in this thread.

jorisvandenbossche replied:
We mainly discussed that for datetime64, where I am fine with that. 
But for Period, you implicitly used the start_times for taking the mean, while you could also have taken the end_times. So I still find this ambiguous. 

(anyway, I think we agree to disallow it all-together for all Periods, so this is only about the error message)

jorisvandenbossche reviewed (changes requested):
To repeat what I said before: to move forward on this, I would suggest to simply remove the axis keyword in this PR. We can always add it later when if we actually want fix the numpy compatibility, or need it for another reason.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jorisvandenbossche commented:
@jbrockmendel instead of keeping rebasing this, can you actually answer to my request for changes? (unless it is just lack of time that you didn't make any changes yet, that's no problem)

jreback commented:
can you merge master and fix the conflicts.

@jorisvandenbossche this has been outstanding a while, any remaining issues?

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jreback commented:
@jbrockmendel also pls respond to open questions from @jorisvandenbossche

jreback commented:
can you merge master; this looks fine otherwise, though I think @jorisvandenbossche had a question.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

TomAugspurger reviewed (approved):
This seems good for now. There's still an outstanding discussion about accepting additional arguments so that methods like `np.mean` work correctly. These are

* axis
* dtype
* out
* keepdims (optional)

but I'm fine with leaving that as a followup when someone wants it.

jorisvandenbossche committed:
remove axis keyword for now

jorisvandenbossche commented:
Given the lack of response of @jbrockmendel, I added a commit that removes the `axis` keyword (single commit, so can be easily removed again if that is a problem). 
Rationale: since we don't do any compat-handling of numpy keyword arguments, let's also not do it for axis. We can always add that later.

Fine with merging it like this.

jreback commented:
ok, pls merge on green. @jorisvandenbossche I guess worthwile to create a followup issue.

jbrockmendel commented:
> Given the lack of response of @jbrockmendel,

My bad.  Starting next month I'm working FT on pandas, so until then I'm focusing on statsmodels and spending time outdoors.  Will rebase momentarily.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel committed:
Merge branch 'means' of https://github.com/jbrockmendel/pandas into means

jreback commented:
I think this will pass if you merge master?

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jorisvandenbossche committed:
don't pass axis to methods

jorisvandenbossche committed:
add returns to docstring

jorisvandenbossche commented:
> until then I'm focusing on statsmodels and spending time outdoors.

Enjoy your time off!

---

Failures were actually real ones: pushed fixes for a problem with the axis (that I caused) and a docstring lint error

jreback commented:
@jbrockmendel one more merge master pls.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into means

jbrockmendel commented:
gentle ping

jreback commented:
thanks @jbrockmendel
